src/content/blog/2023-09-26-preparing-custom-rules-eslint-v9.md

```diff
@@ -0,0 +1,245 @@
+teaser: "The upcoming major release of ESLint includes several breaking changes for rule authors. Read on to learn how to update your rules for ESLint v9.0.0."
+tags:
+  - Rules
+  - API
+authors:
+  - nzakas
+categories:
+  - API Changes
+---
+
+When ESLint v9.0.0 is released, it will ship with several breaking changes for rule authors. These changes are necessary as part of the work to implement [language plugins](https://github.com/eslint/rfcs/blob/main/designs/2022-languages/README.md), which gives ESLint first-class support for linting languages other than JavaScript. We've had to make these changes because ESLint has, from the start, assumed that it would only ever be used to lint JavaScript. As such, there wasn't a lot of thought put into where methods that rules used to interact with source code should live. When revisiting the API for the language plugins work we found that the inconsistencies we were able to live with in a JavaScript-only world will not work in a language-agnostic ESLint core.
+
+## From `context` to `SourceCode`
+
+The majority of the breaking changes for rule authors consist of moving methods off of the `context` object that is passed into rules and onto the `SourceCode` object that is retrieved via `context.sourceCode` (or the deprecated `context.getSourceCode()`; see below). The area of responsibility for `context` vs. `SourceCode` has shifted during the lifetime of ESLint: In the beginning, `context` was the home for everything rules needed to use. Once we added `SourceCode`, we slowly started adding more methods to it. The end result was that some methods lived on `context` and some methods lived on `SourceCode`, and the only reason why? The time at which the methods were added.
+
+In a language-agnostic ESLint core, we need to redefine the responsibilities of these two objects. Going forward, `context` is the home for functionality that rules need to interact with the core while `SourceCode` is the home for functionality that rules need to interact with the code being linted. This allows the same `context` object to be used regardless of the language being linted as well as allowing language plugins to define their own `SourceCode` class to provide methods that are unique to that language.
+
+All of this is to say that we are deprecating all of the code-related methods on `context` and moving them to `SourceCode`. The following table shows which fields on `context` are moving to `SourceCode`. Note that the method signatures remain unchanged for all of these methods even if the name changes:
+
+|**Deprecated on `context`**|**Replacement on `SourceCode`**|
+|-----------------------|--------------------------|
+|`context.getSource()`|`sourceCode.getText()`|
+|`context.getSourceLines()`|`sourceCode.getLines()`|
+|`context.getAllComments()`|`sourceCode.getAllComments()`|
+|`context.getNodeByRangeIndex()`|`sourceCode.getNodeByRangeIndex()`|
+|`context.getComments()`|`sourceCode.getComments()`|
```

Comment:
`sourceCode.getComments()` is actually going to be removed in v9: https://github.com/eslint/eslint/issues/14744.

`sourceCode.getCommentsBefore()`, `sourceCode.getCommentsAfter()`, and `sourceCode.getCommentsInside()` should be used instead.

Reply:
Good catch!